### DISCUSSION TITLE
How to optimize `$queryRaw` to avoid N+1 issues with GraphQL field resolution?

### DISCUSSION THREAD

**Discussion by poteboy** [prisma/prisma#25407] [Q&A] [answered]
### Question

I'm currently working on a project that involves using PostgreSQL with PostGIS, and due to [Prisma's lack of support for PostGIS](https://github.com/prisma/prisma/issues/2789), I have to rely heavily on `$queryRaw` for raw SQL queries. However, I’ve encountered what seems like N+1 issues when querying related fields via GraphQL, specifically when using `$queryRaw` to fetch data.

Here's an example scenario:

I have a GraphQL query where I'm fetching nearby places along with their addresses. I need to calculate the distance using `ST_DistanceSphere` and filter the results by radius using `ST_DWithin`, so I'm using `$queryRaw` for the main query:

```sql
SELECT p.id, p.name, a.address, ST_AsText(a.geometry) as geometry, 
       ST_DistanceSphere(a.geometry, ST_MakePoint(${longitude}, ${latitude})) AS distance
FROM "Place" p
INNER JOIN "Address" a ON a."placeId" = p.id
WHERE ST_DWithin(a.geometry, ST_SetSRID(ST_MakePoint(${longitude}, ${latitude}), 4326), ${radius})
ORDER BY distance
LIMIT ${count};
```

In a separate query, I fetch the Address details:

```sql
SELECT a.id, a."placeId", a.address, ST_AsText(a.geometry) as geometry
FROM "Address" a
WHERE a."placeId" = ${id};
```

When resolving the address field in GraphQL, I issue this query separately for each place. This results in Prisma issuing multiple queries, leading to what looks like an N+1 problem.

## Important note:

I understand that if I were using Prisma’s built-in ORM methods (e.g., `findMany` with `include`), Prisma would automatically batch and optimize these queries via DataLoader, preventing N+1 issues. However, because I’m using `$queryRaw` for PostGIS-related queries, this optimization does not seem to apply.

## My question:

1. Is there a best practice or recommended approach to optimize $queryRaw usage in Prisma to avoid N+1 issues when resolving GraphQL fields?
2. Does Prisma support any form of query batching when using $queryRaw?
3. Is there a way to manually implement DataLoader-style batching for raw SQL queries in Prisma?
4. Alternatively, is there any recommended pattern for working with PostGIS while still leveraging Prisma’s query optimization capabilities?

### Expected behavior (optional)

Ideally, I would expect Prisma to batch or optimize the queries when using `$queryRaw`, similar to how it does with standard `findMany` or `include` methods, to avoid N+1 issues when querying related fields. Alternatively, a best practice for handling this kind of situation when raw queries are required would be very helpful.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model Place {
  id            String          @id @default(cuid())
  name          String
  address       Address?
  createdAt     DateTime         @default(now())
  updatedAt     DateTime         @updatedAt
}

model Address {
  id           String   @id @default(cuid())
  placeId      String?   @unique
  address      String
  geometry     Unsupported("geometry(Point,4326)")
  place        Place?    @relation(fields: [placeId], references: [id])
}
```

```graphql
query NearbyPlaces(
  $latitude: Float!
  $longitude: Float!
  $radius: Float!
  $count: Int
) {
  nearbyPlaces(
    input: { latitude: $latitude, longitude: $longitude, radius: $radius, count: $count }
  ) {
    id
    name
    address {
      geometry
      address
    }
  }
}
```

**Comment by poteboy** [comment, marked as answer]
I’ve resolved the issue by using [DataLoader](https://github.com/graphql/dataloader)

**Reply by kasir-barati** [reply]
Thanks for sharing the result, BTW I am curious as to where you read in the Prisma doc that they optimize their queries when we use their API? I mean I have enabled the logging functionality of `PrismaClient` and I can see that it is sending separate queries. But if they are using Dataloader I would love to know a bit more.

Especially with their [Prisma v6 feature where you can specify that you wanna perform a join at the DB level](https://www.prisma.io/blog/prisma-6-better-performance-more-flexibility-and-type-safe-sql#pick-the-best-join-strategy), thus solving this problem from its root.

**Reply by kasir-barati** [reply]
JFYI: in the Prisma doc for `include` (what they call it "Nested reads") it is written that:

> query: Sends multiple queries to the database (one per table) and joins them on the application level.
> 
> &mdash; [Ref](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#relation-load-strategies-preview).

I cannot see anything related to Dataloader or batching for that matter. Although their new move is somewhat really promising if they manage to pull it off successfully (I mean the join at the DB level).

## Update

I just read a very interesting post about `LATERAL JOIN`s in PSQL which was stating that it is not performant when you have large data sets: https://www.cybertec-postgresql.com/en/understanding-lateral-joins-in-postgresql/#comment-1366
